### PR TITLE
l2geth: Add berlin hardfork

### PR DESCRIPTION
**Pull request by mslipper**

### CODE REVIEW

**Review by smartcontracts — APPROVED** [review]
LGTM. My feedback on this PR can be found in the previous version of this PR located at #1798. We reviewed this PR extensively together on a video chat approximately one week ago.

File: l2geth/params/protocol_params.go [hunk 1]
```diff
@@ -65,23 +69,31 @@ const (
...
+	ExpGas                    uint64 = 10    // Once per EXP instruction
+	LogGas                    uint64 = 375   // Per LOG* operation.
+	CopyGas                   uint64 = 3     //
+	StackLimit                uint64 = 1024  // Maximum size of VM stack allowed.
+	TierStepGas               uint64 = 0     // Once per operation, for a selection of them.
+	LogTopicGas               uint64 = 375   // Multiplied by the * of the LOG*, per LOG transaction. e.g. LOG0 incurs 0 * c_txLogTopicGas, LOG4 incurs 4 * c_txLogTopicGas.
+	CreateGas                 uint64 = 32000 // Once per CREATE operation & contract-creation transaction.
+	Create2Gas                uint64 = 32000 // Once per CREATE2 operation
+	SelfdestructRefundGas     uint64 = 24000 // Refunded following a selfdestruct operation.
+	MemoryGas                 uint64 = 3     // Times the address of the (highest referenced byte in memory + 1). NOTE: referencing happens on read, write and in instructions such as RETURN and CALL.
```
Comment (smartcontracts) [inline l2geth/params/protocol_params.go:92]: Need a newline below this to match upstream geth if we want to avoid unnecessary diff

Comment (smartcontracts) [reply l2geth/params/protocol_params.go:92]: Otherwise this file matches upstream

File: l2geth/params/config.go [hunk 2]
```diff
@@ -215,16 +215,16 @@ var (
 	//
 	// This configuration is intentionally not using keyed fields to force anyone
 	// adding flags to the config to also have to set these fields.
-	AllEthashProtocolChanges = &ChainConfig{big.NewInt(108), big.NewInt(0), nil, false, big.NewInt(0), common.Hash{}, big.NewInt(0), big.NewInt(0), big.NewInt(0), big.NewInt(0), big.NewInt(0), big.NewInt(0), nil, nil, new(EthashConfig), nil}
```
Comment (smartcontracts) [inline l2geth/params/config.go:218]: Matches upstream

File: l2geth/les/api_backend.go [hunk 3]
```diff
@@ -199,7 +199,7 @@ func (b *LesApiBackend) GetTd(hash common.Hash) *big.Int {
 	return b.eth.blockchain.GetTdByHash(hash)
 }
 
-func (b *LesApiBackend) GetEVM(ctx context.Context, msg core.Message, state *state.StateDB, header *types.Header) (*vm.EVM, func() error, error) {
+func (b *LesApiBackend) GetEVM(ctx context.Context, msg core.Message, state *state.StateDB, header *types.Header, vmCfg vm.Config) (*vm.EVM, func() error, error) {
 	state.SetBalance(msg.From(), math.MaxBig256)
```
Comment (smartcontracts) [inline l2geth/les/api_backend.go:203]: To avoid diff with upstream, there's a bug here. The contents should be:

```
	if vmConfig == nil {
		vmConfig = new(vm.Config)
	}
	txContext := core.NewEVMTxContext(msg)
	context := core.NewEVMBlockContext(header, b.eth.blockchain, nil)
	return vm.NewEVM(context, txContext, state, b.eth.chainConfig, *vmConfig), state.Error, nil
```

Comment (smartcontracts) [reply l2geth/les/api_backend.go:203]: And the param should be a pointer

File: l2geth/internal/ethapi/backend.go [hunk 4]
```diff
@@ -59,7 +59,7 @@ type Backend interface {
 	StateAndHeaderByNumberOrHash(ctx context.Context, blockNrOrHash rpc.BlockNumberOrHash) (*state.StateDB, *types.Header, error)
 	GetReceipts(ctx context.Context, hash common.Hash) (types.Receipts, error)
 	GetTd(hash common.Hash) *big.Int
-	GetEVM(ctx context.Context, msg core.Message, state *state.StateDB, header *types.Header) (*vm.EVM, func() error, error)
+	GetEVM(ctx context.Context, msg core.Message, state *state.StateDB, header *types.Header, vmCfg vm.Config) (*vm.EVM, func() error, error)
```
Comment (smartcontracts) [inline l2geth/internal/ethapi/backend.go:62]: In upstream this is a pointer:

```
, vmCfg vm.Config
```

https://github.com/ethereum/go-ethereum/blob/64c53edf836efb1d55b85e841db364a424e95422/internal/ethapi/backend.go#L70

File: l2geth/eth/api_backend.go [hunk 5]
```diff
@@ -271,7 +253,7 @@ func (b *EthAPIBackend) GetTd(blockHash common.Hash) *big.Int {
 	return b.eth.blockchain.GetTdByHash(blockHash)
 }
 
-func (b *EthAPIBackend) GetEVM(ctx context.Context, msg core.Message, state *state.StateDB, header *types.Header) (*vm.EVM, func() error, error) {
+func (b *EthAPIBackend) GetEVM(ctx context.Context, msg core.Message, state *state.StateDB, header *types.Header, vmCfg vm.Config) (*vm.EVM, func() error, error) {
```
Comment (smartcontracts) [inline l2geth/eth/api_backend.go:256]: vm.Config should be a pointer

File: l2geth/eth/api_backend.go [hunk 6]
```diff
@@ -282,7 +264,7 @@ func (b *EthAPIBackend) GetEVM(ctx context.Context, msg core.Message, state *sta
 	vmError := func() error { return nil }
```
Comment (smartcontracts) [inline l2geth/eth/api_backend.go:265]: Need to add:

https://github.com/ethereum/go-ethereum/blob/64c53edf836efb1d55b85e841db364a424e95422/eth/api_backend.go#L206-L208

File: l2geth/les/api_backend.go (hunk 3 above)
Comment (mslipper) [reply l2geth/les/api_backend.go:203]: Our version of geth doesn't use `NewEVMTxContext`, so we can't add these methods. I also don't think we support LES, so I left this as-is.

File: l2geth/params/protocol_params.go (hunk 1 above)
Comment (mslipper) [reply l2geth/params/protocol_params.go:92]: Fixed

Comment (mslipper) [reply l2geth/params/protocol_params.go:92]: Actually, reverted that - would force me to actually increase the diff by running `goimports`
